backend/cellguide/pipeline/ontology_tree/tree_builder.py

```diff
@@ -205,3 +213,332 @@ def _traverse_ontology_with_counting(
+                )
+
+                all_paths = []
+                siblings = []
+                for i in range(self.traverse_node_counter[end_node]):
+                    end_node_i = end_node + "__" + str(i)
+                    path = self._depth_first_search_pathfinder(end_node_i)
+                    path = path if path else [end_node_i]
+                    all_paths.append(path)
+
+                    siblings.append(
+                        sum([self.all_children.get(parent, []) for parent in self.all_parents.get(end_node_i, [])], [])
+                    )
+
+                visited_nodes_in_paths = list(set(sum(all_paths, [])))  # in a path to target
+
+                children = self.all_children.get(end_node + "__0", [])  # children
+                grandchildren = sum([self.all_children.get(child, []) for child in children], [])  # grandchildren
+                siblings = list(set(sum(siblings, [])))  # siblings
+                valid_nodes = list(set(visited_nodes_in_paths + children + grandchildren + siblings))
+
+                ontology_graph_copy = self._get_deepcopy_of_ontology_graph()
+                self._truncate_graph_first_pass(ontology_graph_copy, valid_nodes)
+                self._truncate_graph_second_pass(ontology_graph_copy, visited_nodes_in_paths)
+
+                isExpandedNodes = list(set(_getExpandedData(ontology_graph_copy)))
+                notShownWhenExpandedNodes = _getShownData(ontology_graph_copy)
+
+                notShownWhenExpanded = {}
+                for i in notShownWhenExpandedNodes:
+                    notShownWhenExpanded.update(i)
+                all_states_per_cell_type[end_node] = {
+                    "isExpandedNodes": isExpandedNodes,
+                    "notShownWhenExpandedNodes": notShownWhenExpanded,
+                }
+
+        return all_states_per_cell_type
+
+    def _depth_first_search_pathfinder(self, path_end_node, node=None, path=None):
+        """
+        This function finds a path backwards from the end node to the start node using depth-first search.
+
+        Arguments
+        ---------
+        path_end_node: str
+            The end node of the path.
+        node: str, optional, default=None
+            The current node in the depth-first search.
+        path: list, optional, default=None
+            The current path in the depth-first search.
+
+        Returns
+        -------
+        path: list
+            The path from the end node to the start node.
+        """
+        if path is None:
+            path = [path_end_node]
+            node = path_end_node
+
+        if node == self.start_node:
+            return path
+
+        for parent in self.all_parents.get(node, []):
+            full_path = self._depth_first_search_pathfinder(path_end_node, node=parent, path=path + [parent])
+            if full_path:
+                return full_path
+
+    def _truncate_graph_first_pass(self, graph, valid_nodes):
```

Comment:
you can have this function return nothing. just update the signature to be
```
def _truncate_graph_first_pass(self, graph, valid_nodes) -> None:
```
and remove the line at the bottom that says `return True`